main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -18,264 +18,564 @@ package ca.uwaterloo.flix.language.phase
+      case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.SymmetricDiff, _), x, _), y, _) =>
+        eval(x).flatMap(a => eval(y).map(b => CofiniteEffSet.xor(a, b)))
+      case Type.Alias(_, _, tpe, _) => eval(tpe)
+      case Type.Var(_, _) => throw InternalCompilerException(s"Unexpected effect '$eff'.", eff.loc)
+      case Type.Apply(_, _, _) => throw InternalCompilerException(s"Unexpected effect '$eff'.", eff.loc)
+      case Type.AssocType(_, _, _, _) => throw InternalCompilerException(s"Unexpected effect '$eff'.", eff.loc)
+      case Type.JvmToType(_, _) => throw InternalCompilerException(s"Unexpected effect '$eff'.", eff.loc)
+      case Type.JvmToEff(_, _) => throw InternalCompilerException(s"Unexpected effect '$eff'.", eff.loc)
+      case Type.UnresolvedJvmType(_, _) => throw InternalCompilerException(s"Unexpected effect '$eff'.", eff.loc)
+    }
 
-  /**
-    * Returns `true` iff `declaredEff` is not `Pure` and contains some non-primitive effect.
-    */
-  private def isBadEntryPointEffect(declaredEff: Type) = {
-    declaredEff.effects.exists(sym => !Symbol.isPrimitiveEff(sym))
-  }
+    /** Returns an error if `defn` is in the root namespace. */
+    private def checkNonRootNamespace(defn: TypedAst.Def): Option[EntryPointError] = {
+      val inRoot = defn.sym.namespace.isEmpty
+      if (inRoot) Some(EntryPointError.IllegalExportNamespace(defn.sym.loc))
+      else None
+    }
+
+    /** Returns an error if `defn` is not a public function. */
+    private def checkPub(defn: TypedAst.Def): Option[EntryPointError] = {
+      val isPub = defn.spec.mod.isPublic
+      if (isPub) None
+      else Some(EntryPointError.NonPublicExport(defn.sym.loc))
+    }
+
+    /** Returns `None` if `defn` has a name that is valid in Java. Returns an error otherwise. */
+    private def checkValidJavaName(defn: TypedAst.Def): Option[EntryPointError] = {
+      val validName = defn.sym.name.matches("[a-z][a-zA-Z0-9]*")
+      if (validName) None
+      else Some(EntryPointError.IllegalExportName(defn.sym.loc))
+    }
+
+    /**
+      * Returns an error for each type in `defn` that is not valid in Java.
+      *
+      * N.B.: `defn` must not have type variables or associated types.
+      */
+    private def checkJavaTypes(defn: TypedAst.Def): List[EntryPointError] = {
+      val types = defn.spec.retTpe :: defn.spec.fparams.map(_.tpe)
+      types.flatMap(tpe => {
+        if (isJavaType(tpe)) None
+        else Some(EntryPointError.IllegalExportType(tpe, tpe.loc))
+      })
+    }
+
+    /**
+      * Returns `true` if `tpe` is a valid Java type.
+      *
+      *   - `isJavaType(Int32) = true`
+      *   - `isJavaType(Bool) = true`
+      *   - `isJavaType(String) = true`
+      *   - `isJavaType(List[String]) = false`
+      *   - `isJavaType(java.lang.Object) = true`
+      *
+      * N.B.: `tpe` must not have type variables or associated types.
+      */
+    @tailrec
+    private def isJavaType(tpe: Type): Boolean = {
```

Comment:
Does this belong here?

Reply:
For now, since its in a limited state, i think it should live here